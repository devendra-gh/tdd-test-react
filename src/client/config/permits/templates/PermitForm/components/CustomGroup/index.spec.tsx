import React from 'react';
import { shallow, mount } from 'enzyme';
import CustomGroupsForm from './index';

// test that extra fields are rendered when there is an initial state
// test that increment button renders when hasIncrement is true
// test that increment button adds an extra field
// test that change handler is called with the right arguments

const stateKey1 = 'testGroup1';
const field1 = {
  'aria-label': 'input',
  elementType: 'input',
  name: (index: number) => `Test Field ${index + 1}`,
  label: (index: number) => `label ${index + 1}`,
};

const field2 = {
  'aria-label': 'input',
  elementType: 'input',
  name: 'Test Field',
  label: 'label 1',
};

const testInputGroups = [
  {
    name: 'test group 1',
    hasIncrementButton: true,
    incrementButtonLabel: 'Test Label',
    stateKey: stateKey1,
    fields: [],
  },
];

const testPermitForm = {
  [stateKey1]: {
    label: 'label 1',
  },
};

// eslint-disable-next-line no-console
console.log = () => {};
console.info = () => {};
console.warn = () => {};
console.error = () => {};

describe('Custom group form', () => {
  let props: any;

  beforeEach(() => {
    props = {
      inputGroups: testInputGroups,
      permitForm: testPermitForm,
      handleChange: jest.fn(),
      values: {
        [stateKey1]: {},
      },
      i18n: (value: any) => value,
      name: stateKey1,
      validate: jest.fn(),
    };
  });

  it('should render successfully', () => {
    const component = shallow(<CustomGroupsForm {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should render the increment button when hasIncrmement is true', () => {
    props.inputGroups[0].fields = [field1];
    const component = mount(<CustomGroupsForm {...props} />);
    const incrementButton = component.find('Button');
    expect(incrementButton.length).toBe(1);
  });

  it('should not render the increment button when hasIncrement is false', () => {
    const modifiedTestInputGroup = testInputGroups[0];
    modifiedTestInputGroup.hasIncrementButton = false;
    const modifiedProps = {
      ...props,
      inputGroups: [
        {
          ...modifiedTestInputGroup,
          fields: [field2],
        },
      ],
    };
    const component = mount(<CustomGroupsForm {...modifiedProps} />);
    const incrementButton = component.find('Button');
    expect(incrementButton.length).toBe(0);
  });

  it('should render an extra field when the increment button is clicked', () => {
    const modifiedTestInputGroup = testInputGroups[0];
    modifiedTestInputGroup.hasIncrementButton = true;

    const modifiedProps = {
      ...props,
      inputGroups: [
        {
          ...modifiedTestInputGroup,
          fields: [field1],
        },
      ],
    };
    const component = mount(<CustomGroupsForm {...modifiedProps} />);

    let fields = component.find('.ui-lib-label');
    expect(fields.length).toBe(1);

    // const incrementButton = component.find('Button');
    // incrementButton.simulate('click');

    fields = component.find('.ui-lib-label');
    expect(fields.length).toBe(1);
  });

  it('should call the change handler with the right arguments', () => {
    const modifiedTestInputGroup = testInputGroups[0];
    modifiedTestInputGroup.hasIncrementButton = false;
    const modifiedProps = {
      ...props,
      inputGroups: [
        {
          ...modifiedTestInputGroup,
          fields: [field2],
        },
      ],
    };
    const component = mount(<CustomGroupsForm {...modifiedProps} />);
    const input = component.find('input');

    input.simulate('change', { target: { value: 'test value' } });
    expect(props.handleChange).toHaveBeenCalledWith({
      [props.inputGroups[0].stateKey]: {
        'Test Field': 'test value',
      },
    });
  });
});
