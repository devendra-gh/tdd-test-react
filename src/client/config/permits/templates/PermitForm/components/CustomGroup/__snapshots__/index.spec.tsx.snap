// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom group form should render successfully 1`] = `
<FormGroup
  component={[Function]}
  fields={Array []}
  formProps={
    Object {
      "handleChange": [MockFunction],
      "i18n": [Function],
      "inputGroups": Array [
        Object {
          "fields": Array [],
          "hasIncrementButton": true,
          "incrementButtonLabel": "Test Label",
          "name": "test group 1",
          "stateKey": "testGroup1",
        },
      ],
      "name": "testGroup1",
      "permitForm": Object {
        "testGroup1": Object {
          "label": "label 1",
        },
      },
      "validate": [MockFunction],
      "values": Object {
        "testGroup1": Object {},
      },
    }
  }
  handleChange={[MockFunction]}
  hasIncrementButton={true}
  i18n={[Function]}
  incrementButtonLabel="Test Label"
  key="testGroup1"
  name="test group 1"
  stateKey="testGroup1"
  validate={[MockFunction]}
  values={
    Object {
      "testGroup1": Object {},
    }
  }
/>
`;
